### DISCUSSION TITLE
Dietpi-Software | Plexamp headless

### DISCUSSION THREAD

**Discussion by tgp-2** [MichaIng/DietPi#6048] [Software Request]
#### Formal software information
- Software title | Plexamp headless
- Short description | use device as a Plexamp audio streaming endpoint
- Official URL (if available) | https://www.plex.tv/plexamp/
- GitHub URL (if available) | n/a
- (Official) install documentation | https://forums.plex.tv/t/plexamp-on-the-raspberry-pi/791500

#### Are there similar/alternative software titles available with DietPi-Software?
Plex Media Server (related server app, not a playback app)

#### What makes your requested software better than the above solutions, if available?
Plexamp headless is a dedicated remote controllable music playback app, for use together with Plex Media Server (preferably running on a different device) and optional paid Tidal subscription

#### How can DietPi make the installation easier or compatible, than following the install instructions or do APT installation, if available?
current instructions are general and assume Linux terminal knowledge, requires user to install/update dependencies (such as nodejs) and edit systemd service files ... Plex reps might make install easier in the future, but as of now no details/timeline have been provided ... there are community supported installers, for example [this one](https://github.com/odinb/bash-plexamp-installer) already provides best effort support for install on DietPi, but supporting in `dietpi-software` might make it easier for users to install and use alongside other audio/streaming apps such as `shairport-sync` or `roonbridge`

#### Can you provide the installation steps that you would suggest DietPi-Software to do?
1. confirm user to use for install and systemd service (default: dietpi), add to audio group as needed
2. (optional ... not required for install, but helpful) show list of audio devices, have user confirm desired device (HAT, USB DAC, etc.) is available
3. install nodejs (currently Plexamp headless requires node 16) ... easiest to install DietPi system version if this matches what is required by Plexamp, but required version for Plexamp might change out of sync with DietPi ... in which case it's also possible to use nvm and keep a separate version of nodejs just for Plexamp (example script below gives example of an nvm-based install)
4. install other dependencies as needed (jq, bzip2, wget, tar)
5. download and extract target version of Plexamp headless in user home directory 
6. prepare systemd service files (edit defaults provided by Plex)
7. start initial Plexamp instance, user will be asked to copy/paste Plex claim token (initial install only) ... note, as of version 4.7.0, claim token can also be entered using environment variable (in case this makes DietPi-Software install easier to implement) 
8. exit installer and start/restart Plexamp service as needed
9. instruct user to complete app configuration at `<hostname or ip>:32500` 

example bash install script using nvm (this example assumes script run by non-root user in sudo group) ...

```
#! /bin/bash
# install Plexamp headless on a Raspberry Pi or similar system
# uses nvm to install/manage node version, Plexamp currently requires node 16
# usage: from user home directory, bash ./plexamp-install.sh
#
# install nvm
cd $HOME
if [ ! $HOME/.nvm ]; then
  mkdir $HOME/.nvm
fi
export NVM_DIR="$HOME/.nvm"
wget -qO- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh | bash
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
# install or update node 16
nvm install 16
NODE=$(node -v)
# bzip2 jq for current install and future upgrades
sudo apt install jq bzip2
# backup current
cd $HOME
rm -rf plexamp.last
mv -f plexamp plexamp.last
# download and extract Plexamp archive
# hardcoded version below, will update as needed
wget -O plexamp-temp.tar.bz2 https://plexamp.plex.tv/headless/Plexamp-Linux-headless-v4.7.4.tar.bz2
tar -xjvf plexamp-temp.tar.bz2
rm plexamp-temp.tar.bz2
# prepare Plexamp systemd service
cd $HOME/plexamp
sed -i "s/User=pi/User=$USER/" plexamp.service
sed -i "s/home\/pi/home\/$USER/" plexamp.service
sed -i "s/usr\/bin\/node/home\/$USER\/.nvm\/versions\/node\/$NODE\/bin\/node/" plexamp.service
sudo cp plexamp.service /lib/systemd/system/
sudo systemctl daemon-reload
sudo systemctl enable plexamp.service
# start initial Plexamp session
node js/index.js
```

other relevant notes/info ...

[requires]
- Plex Media Server running elsewhere and a Plex Pass subscription
- system with a compatible ARMv7/v8 or x86_64 CPU
- audio device attached/enabled (on-board audio, USB audio, HAT, etc.)
- DietPi OS based on Debian Bullseye or later (as of May '23, glibc >= 2.31)
- recommend system with >= 1GB DRAM for best performance (512MB minimum)

[app install]
- as of May '23, Plexamp execution requires nodejs version 16
- nodejs install/updates can use nvm (allows Plexamp nodejs version <> system) 
- install and updates require jq, bzip2, wget, tar
- initial install on a specific device requires Plex claim token entry, either via interactive shell or environment variable 
- Plexamp runs as non-root user with app/config/log files in user home directory 
- install configures systemd system service for Plexamp autostart

[app updates]
- Plex provided update script used for most app updates (`~/plexamp/upgrade.sh`)
- run latest installer if updates to nodejs (and/or nvm) are also needed

[app configuration and usage notes]
- adjust audio device gain setting as needed using alsamixer (Plexamp does not touch system/ALSA settings)
- hostname or ip:32500 for Plexamp device specific settings
- playback --> select audio device + desired settings for sample rate matching, volume leveling, preamp gain, and limiter
- advanced --> select desired settings for caching, etc. (recommend lowest 512MB cache size and Wi-Fi caching set to "Only next track")
- note: when Plexamp headless playback is paused/stopped, the audio device will be released after a short delay (about 10 sec)

[Plexamp log files]
- `~/.cache/Plexamp/log`

#### Are you willing to help maintaining the software installation, e.g. in case of needed setup changes due to updates etc.? This is not needed, but could speed up our decision to implement it, as man power is always a topic :wink:.
yes, I am willing and currently available to help maintain, but my bash skills are limited

**Comment by Rickyrivits** [comment]
Did anyone ever build Dietpi OS application(s) for "Plexamp?Looking for a headless playback endpoint streamer (NOT looking for a PLEX Server app) Currently using Dietpi with ROON as a endpoint too. So something like it for Plexamp.
I understand today I can build headless streamer using raspberry PI os. I have this build now, So it would be nice to be able to use both Roon and Plexamp on the same RPI.
